### PR TITLE
Permit empty reason strings in ClientResponse.raise_for_status()

### LINKED ISSUE #3532: ClientResponse.raise_for_status() raises an AssertionError for empty reason
## Long story short

`ClientResponse.raise_for_status()` raises an `AssertionError` for a >= 400 status code if its `reason` is an empty string.

## Expected behaviour

I expect `ClientResponse.raise_for_status()` to raise a `ClientResponseError` if it’s invoked on a `ClientResponse` with a status >= 400, and to either:

* raise a `ClientError` (or some subclass of `ClientError`) if it’s called on a `ClientResponse` with a `reason = None`, regardless of the HTTP status code, if this can only happen when `raise_for_status()` has been called before `start()` has been called; or
* convert a `ClientResponse.reason` of `None` into `""` if `None` is a valid value for `ClientResponse.reason`

## Actual behaviour

* `ClientResponse.raise_for_status()` raises an AssertionError if it’s invoked on a `ClientResponse` when a status >= 400 and `bool(reason)` is `False`, which includes `reason == ""`
* `ClientResponse.raise_for_status()` raises a ClientResponseError if it’s invoked on a `ClientResponse` with a status >= 400 if the `reason` is not falsey
* `ClientResponse.raise_for_status()` raises no error if it’s invoked on a `ClientResponse` with a status < 400 and a `reason = None`

## Steps to reproduce

A colleague and I uncovered this using [aioresponses](https://github.com/pnuckowski/aioresponses), which currently doesn’t set a `reason` on the `ClientResponse` at all – see https://github.com/pnuckowski/aioresponses/issues/114.

You can reproduce this with the example client and server in this gist: https://gist.github.com/rhwlo/5ff3cb6bae8deb6abb4a1e527ad9583e

## Your environment

I’m using the latest `aiohttp` from the repository, but this problem shows up only in `v3.5.0` and newer.

### PR DESCRIPTION
**Pull request by rhwlo**
## What do these changes do?

* Allows `ClientResponse.raise_for_status` to raise a `ClientResponseError` in the case of an HTTP Reason-Phrase that is an empty string
* Raises a `TypeError` if the `ClientResponse.reason` is `None` (since we can’t make a `ClientResponseError` with a message of `None`)
* Fixes https://github.com/aio-libs/aiohttp/issues/3532

## Are there changes in behavior for the user?

* Calling `raise_for_status` on a status with a reason phrase that’s an empty string will raise a `ClientResponseError` now instead of an `AssertionError`

## Related issue number

https://github.com/aio-libs/aiohttp/issues/3532

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: aiohttp/client_reqrep.py [hunk 1]
```diff
@@ -931,8 +931,10 @@ def release(self) -> Any:
         return noop()
 
     def raise_for_status(self) -> None:
+        # self.reason is only None if self.start() hasn't been called:
+        if self.reason is None:
+            raise TypeError("Expected a Reason-Phrase as a string")
```
Comment (asvetlov) [inline aiohttp/client_reqrep.py]: An exception type and especially text is meaningless for an end user.
The text describes a consequence (the `reason` field is not set) but hides an actual problem.

Moreover, assertion exists only for sake of `mypy`.
If you don't like the assertion -- please find another way to satisfy mypy.

`assert self.reason is not None` could be the simplest possible fix, even no test is needed for the case.

Comment (asvetlov) [reply aiohttp/client_reqrep.py]: Well, a test would be nice but please use a functional test for an empty `reason`.
https://github.com/aio-libs/aiohttp/blob/master/tests/test_client_functional.py can give an inspiration how a functional client test looks like

Comment (rhwlo) [reply aiohttp/client_reqrep.py]: I’ll switch it to `assert self.reason is not None` if that’s the desired outcome.

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks!

**Comment by ctg3** [comment]
Why is reason string required? I have a webserver that doesn't return any reason phrase (None). If self.reason is None, can we just change it to self.reason = '' and still satisfy mypy?

Also see a discussion here: https://stackoverflow.com/questions/17517086/can-an-http-response-omit-the-reason-phrase

**Comment by vaultah** [comment]
Maybe make a release that contains a fix for this? No hurry though, it's only been three months.

**Comment by Solvero** [comment]
I have encountered this problem after aiohttp upgrade to 3.5.4. My server doesn't return Response-Reason too. HTTP/1.1 [specification](https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html#sec6.1.1) says that:

> The Status-Code is intended for use by automata and the Reason-Phrase is intended **for the human user**. The **client is not required to examine** or display the Reason- Phrase. 

Please see also:
https://stackoverflow.com/questions/17517086/can-an-http-response-omit-the-reason-phrase

**Comment by gwerbin** [comment]
I appreciate all the work the contributors are doing on this library, but it would have been nice to merge and release this as a 3.5.5 "patch". It looks like you're waiting for 3.6 release or something like that?

In the meantime, I'm using my own implementation of this, basically copy-and-pasted from the source code:
```python
def raise_for_status(response: aiohttp.ClientResponse) -> None:
    if 400 <= response.status:
        response.release()
        raise aiohttp.ClientResponseError(
            response.request_info,
            response.history,
            status=response.status,
            message=response.reason or '',
            headers=response.headers
        )
```

**Comment by webknjaz** [comment]
@gwerbin `master` branch is for 4.0. If you want, you can submit a backport PR against `3.5` branch.

**Comment by gwerbin** [comment]
Thanks @webknjaz, My one concern is about this line:

> reason should always be not None for a started response

Does some other part of the application or test suite depend on this being true? I don't want to break anything else with my quick `or ''` fix.

Alternatively, why would reason be `None` here, rather than defaulting to an empty string? I feel like fixing this at the `raise_for_status` point doesn't actually address the problem, that reason can be `None` in the first place instead of an empty string.

**Comment by webknjaz** [comment]
@gwerbin you should normally cherry-pick the entire PR and not try to rip off some parts of it. This includes tests.

**Comment by fiendish** [comment]
raise_for_status on a response object is still giving AssertionError instead of ClientResponseError in the latest release?

**Pull request merged**